### ISSUE TITLE
boost: --with-mpi may build with single-threaded variant of libs

### ISSUE THREAD

**Issue by aw1621107** [Homebrew/homebrew-core#5820]
Or at least it did on my machine. I just deleted the check in the `boost` formula, then ran `brew install boost --with-mpi`. Compilation succeeded on my machine (macOS 10.12, Xcode 8).

Can anyone else replicate this?

**Comment by MikeMcQuaid** [comment]
To help us debug this issue can you explain:
- What you were trying to do (and why)
- What happened
- What you expected to happen
- Step-by-step reproduction instructions (with as minimal input data as possible)

**Comment by aw1621107** [comment]
Originally, I had Boost installed as `boost --c++11 --with-mpi --without-single`. I was trying to install `asio --c++11 --with-boost-coroutine`, but the build failed at the final link step because asio appears to require linking against the single-threaded version of Boost coroutine, which was not built.

The Boost formula prevents `--with-mpi` being used when building both single- and multi-threaded versions, and points to a bug filed with Boost ~3 years ago. I didn't want to drop the `--with-mpi` flag (because I'm one of those people that likes to install `--with-everything`), so I tried deleting the check just to see what would happen.

Instead of the build failing, which I expected to happen due to the presence of the check, the build succeeded, and `asio` built fine after that.

I don't know about reproduction, because I don't know when the check became unnecessary, but this is the approximate order I did things:

  1. `brew install boost --c++11 --with-mpi`. Fails, stating that building both single- and multi-threaded libs `--with-mpi` is not supported.
  2. `brew install boost --c++11 --with-mpi --without-single`. Succeeds.
  3. `brew install asio --c++11 --with-boost-coroutine`. Build proceeds to the end, when the linking step fails due to being unable to find libboost_coroutine (the multi-threaded version is libboost_coroutine_mt)
  4. Delete lines 42-49 in `boost.rb`. This should correspond to the section:
```
    # https://svn.boost.org/trac/boost/ticket/8841
    if build.with?("mpi") && build.with?("single")
      raise <<-EOS.undent
        Building MPI support for both single and multi-threaded flavors
        is not supported.  Please use "--with-mpi" together with
        "--without-single".
      EOS
    end
```
  5. `brew uninstall boost`
  6. `brew install boost --c++11 --with-mpi`. Build succeeds, which is not expected given the check
  7. `brew install asio --c++11 --with-boost-coroutine`. Succeeds

**Comment by MikeMcQuaid** [comment]
@aw1621107 Thanks! Sounds like you've figured out some improvements. Could you try and open a pull request? [This document](https://github.com/Homebrew/brew/blob/master/docs/How-To-Open-a-Homebrew-Pull-Request-(and-get-it-merged).md#how-to-open-a-homebrew-pull-request-and-get-it-merged) should help and I'm happy to walk you through anything else.

Thanks!

**Comment by dunn** [comment]
Can this be closed since #5853 was merged?
